### PR TITLE
Add a torch/parl version of DDPG

### PR DESCRIPTION
**Pull request by heine-er**
##Pull Request Readiness Checklist

Add a torch/parl version of DDPG algorithm.
Added files:
benchmark/torch/ddpg contains: mujoco_model.py, mujoco_agent.py, train.py, .benchmark/train&eval results.png
parl/algorithms/torch adds: ddpg.py

### CODE REVIEW

File: benchmark/torch/ddpg/mujoco_model.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import torch.nn.functional as F
+'''
+Model of DDPG:  defines an Actor/policy network given obs as input,
+                      & a Critic/value network given obs and action as input.
+
+Attributions:  policy(obs):        return action
+                value(obs, action): return value
+                get_actor_params()
+                get_critic_params()
+'''
```
Comment (rical730) [inline benchmark/torch/ddpg/mujoco_model.py:27]: please remove line 23 to line 26, it will be more clear to add description(input & output) under each function.

File: benchmark/torch/ddpg/train.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import argparse
+from parl.utils import logger, tensorboard, ReplayMemory
+from mujoco_model import MujocoModel
+from mujoco_agent import MujocoAgent
+from parl.algorithms import DDPG
+
+
+# Runs policy for X episodes and returns average reward
+# A fixed seed is used for the eval environment
+def eval_policy(agent, env_name, seed, eval_episodes):
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:28]: eval_episodes=5 by default

File: benchmark/torch/ddpg/train.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            state, reward, done, _ = eval_env.step(action)
+            avg_reward += reward
+    avg_reward /= eval_episodes
+    return avg_reward
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--policy", default="DDPG")  # Policy name
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:46]: please remove this line, this is a DDPG example

File: benchmark/torch/ddpg/train.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    parser.add_argument(
+        "--actor_lr", default=1e-3)  # Learning rate of actor network
+    parser.add_argument(
+        "--critic_lr", default=1e-3)  # Learning rate of critic network
+    parser.add_argument(
+        "--policy_freq", default=1,
+        type=int)  # Frequency to train actor and update params
+    args = parser.parse_args()
+
+    print("---------------------------------------")
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:78]: `print` --> `logger.info`

File: benchmark/torch/ddpg/train.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    logger.info('Policy: {}, Env: {}, Seed: {}'.format(args.policy, args.env,
+                                                       args.seed))
+    print("---------------------------------------")
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:81]: `print` --> `logger.info`

File: benchmark/torch/ddpg/train.py [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+        # Train agent after collecting sufficient data
+        if t >= args.start_timesteps:
+            batch_obs, batch_action, batch_reward, batch_next_obs, batch_terminal = replay_memory.sample_batch(
+                args.batch_size)
+            agent.learn(batch_obs, batch_action, batch_reward, batch_next_obs,
+                        batch_terminal)
+
+        if done:
+            # +1 to account for 0 indexing. +0 on ep_timesteps since it will increment +1 even if done=True
```
Comment (rical730) [inline benchmark/torch/ddpg/train.py:143]: You can remove this comment since the log print is clear enough

File: parl/algorithms/torch/ddpg.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+from copy import deepcopy
+import parl
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+'''
+Algorithm of DDPG: describes the mechanism of updating parameters in Model
+
+Contributions: predict(obs): return act         # choose action based on actor net
+               learn(replay_memory_batch)       # 
+               sync_target()                    # update networks
```
Comment (rical730) [inline parl/algorithms/torch/ddpg.py:26]: please move function description under each function

File: parl/algorithms/torch/ddpg.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+            self,
+            model,
+            max_action,
+            discount=None,
+            tau=None,
+            actor_lr=None,
+            critic_lr=None,
+            decay=None,
+            policy_freq=1,
+    ):  # Frequency of delayed policy updates
```
Comment (rical730) [inline parl/algorithms/torch/ddpg.py:42]: please move this comment to line 53

File: benchmark/torch/ddpg/README.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
+## Reproduce DDPG with PARL
+Based on PARL, the DDPG algorithm of deep reinforcement learning has been reproduced, reaching the same level of indicators as the paper in Mujoco benchmarks.
+
+> **Continuous control with Deep Reinforcement Learning**
+>
+> [Continuous control with Deep Reinforcement Learning](https://arxiv.org/abs/1509.02971)
```
Comment (rical730) [inline benchmark/torch/ddpg/README.md:6]: line4 to line6 -->

> Paper: DDPG in [Continuous control with Deep Reinforcement Learning](https://arxiv.org/abs/1509.02971)

File: benchmark/torch/ddpg/README.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+About DDPG:
+
++ A model-free, off-policy algorithm.
++ Can only be used for environment with continuous action spaces.
```
Comment (rical730) [inline benchmark/torch/ddpg/README.md:11]: only ?

File: benchmark/torch/ddpg/README.md [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+### Mujoco games introduction
+
+Please see [here](https://github.com/openai/mujoco-py) to know more about Mujoco games.
+
+### Benchmark result
+
+<img src=".benchmark/train.png" width = "800" height ="300" alt="Performance" />
+
+<img src=".benchmark/evaluation.png" width = "800" height ="300" alt="Performance" />
```
Comment (rical730) [inline benchmark/torch/ddpg/README.md:34]: we only showing the results of the evaluation

File: benchmark/torch/ddpg/README.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
++ gym
++ torch
++ mujoco-py>=1.50.1.0
+
+### Start Training:
+```
+# To train an agent for HalfCheetah-v1 game
+python train.py
+
+# To train for different game and different loss type
```
Comment (rical730) [inline benchmark/torch/ddpg/README.md:49]: different loss type ?

File: benchmark/torch/ddpg/README.md [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+Tricks:
+
++ Replay memory
+
++ Target network: **soft** target updates
+
++ **Batch normalization**(Not implement)
+
++ Noise for exploration
```
Comment (rical730) [inline benchmark/torch/ddpg/README.md:22]: line 8 to line 22 is not necessary
